content/posts/a_confusing_build_failure.md

```diff
@@ -0,0 +1,64 @@
+test.
+
+## Mockito cannot mock final classes
+
+So why was the build failing? When the unit tests ran some of them would fail with an error stating that Mockito(more specifically [mockito-kotlin](https://github.com/nhaarman/mockito-kotlin)) - a library commonly used in Java and Android to provide mock test doubles and for behavior verification - cannot mock final classes.
+
+In order to understand the error let's try to understand first how Mockito works. It creates at runtime objects that extend interfaces or a classes which can be used to verify that some methods were called with specific parameters. If a class that is attempted to be mocked is declared as `final` then the error above will be thrown, as a final class cannot be extended at compile or run time. In a unit test a Kotlin class was mocked and as Kotlin classes are by default
+`final` unless explicitly declared `open`, the error above was thrown.
+
+The culprit was found but we still have not figured out why the tests pass locally but fail on the CI. 
+
+## Maybe Mockito can mock final classes after all
+
+If Mockito cannot mock final classes then why do local tests pass? Well sure there must be some way to mock a final class, and that's exactly what happened by adding in the gradle build file the `org.mockito:mockito-inline` dependency. 
+
+The problem is that the failing module did not compile with the dependency above. Another submodule did compile it, but it was declared on its `build.gradle` as `testImplementation` not `testApi`, meaning that final classes could be mocked on that submodule only, not on the failing module.
+
+On the CI server the build rightly failed, but locally mockito-inline was compiled on the module above(where it shouldn't) and it wrongfully passed. How can the CI server be compiled and tested in a different way than locally? It turns out by running different commands that compile and test the source code. On the CI server the project is compiled and tested with gradle commands, thus the mockito-inline is not compiled and the test that mocks Kotlin classes fails. On the other hand the
+local builds used the test runner that is built in the Android Studio, which happens to compile all libraries irrespectively of how they are declared on each module's `gradle.build`.
+
+## But why did mockito-inline was used on the first place?
+
+The mystery has at last been solved but let's dig a little bit more. Why did a test try to mock an otherwise concrete class? And when/why was the mockito-inline introduced first?
```

Comment:
```suggestion
The mystery has at last been solved but let's dig a little bit more. Why did a test try to mock an otherwise concrete class? And when/why was the mockito-inline dependency introduced first?
```